src/overlays/actors/ovl_Bg_Hidan_Dalm/z_bg_hidan_dalm.c

```diff
@@ -21,17 +29,175 @@ const ActorInit Bg_Hidan_Dalm_InitVars = {
+    }
+}
+#else
 #pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Hidan_Dalm/BgHidanDalm_Init.s")
+#endif
+
+void BgHidanDalm_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgHidanDalm* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Hidan_Dalm/BgHidanDalm_Destroy.s")
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+    Collider_DestroyTris(globalCtx, &this->unk_16C);
+}
+
+void BgHidanDalm_WaitInteraction(BgHidanDalm* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if ((this->unk_16C.base.acFlags & 2) && !func_8008E988(globalCtx) &&
+        (player->swordAnimation == 22 || player->swordAnimation == 23)) {
+        this->unk_16C.base.acFlags &= ~(1 << 1);
+        if (this->unk_16C.list[0].body.bumperFlags & 2 || this->unk_16C.list[1].body.bumperFlags & 2) {
+            this->dyna.actor.posRot.rot.y -= 0x4000;
+        } else {
+            this->dyna.actor.posRot.rot.y += 0x4000;
+        }
+        this->dyna.actor.posRot.pos.x += 32.5f * Math_Sins(this->dyna.actor.posRot.rot.y);
+        this->dyna.actor.posRot.pos.z += 32.5f * Math_Coss(this->dyna.actor.posRot.rot.y);
+
+        func_8002DF54(globalCtx, &this->dyna.actor, 8);
+        this->dyna.actor.flags |= 1 << 4;
+        this->actionFunc = &BgHidanDalm_Shrink;
+        this->dyna.actor.bgCheckFlags &= ~(1 << 1);
+        this->dyna.actor.bgCheckFlags &= ~(1 << 3);
+        this->dyna.actor.speedXZ = 10.0f;
+        Flags_SetSwitch(globalCtx, this->switchFlag);
+        func_8002F7DC(&PLAYER->actor, NA_SE_IT_HAMMER_HIT);
+        Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_DARUMA_VANISH);
+    } else {
+        CollisionCheck_SetAC(globalCtx, &globalCtx->colChkCtx, &this->unk_16C.base);
+    }
+}
+
+void BgHidanDalm_Shrink(BgHidanDalm* this, GlobalContext* globalCtx) {
+    s32 i;
+    Vec3f unkDir;
+    Vec3f unkPos;
+
+    if (Math_ApproxF(&this->dyna.actor.scale.x, 0.0f, 0.004f)) {
+        func_8002DF54(globalCtx, &this->dyna.actor, 7);
+        Actor_Kill(&this->dyna.actor);
+    }
+
+    this->dyna.actor.scale.y = this->dyna.actor.scale.z = this->dyna.actor.scale.x;
+
+    unkPos.x = this->dyna.actor.posRot.pos.x;
+    unkPos.y = this->dyna.actor.posRot.pos.y + this->dyna.actor.scale.x * 160.0f;
+    unkPos.z = this->dyna.actor.posRot.pos.z;
+
+    for (i = 0; i != 4; i++) {
+        unkDir.x = 5.0f * Math_Sins(this->dyna.actor.posRot.rot.y + 0x8000) + (Math_Rand_ZeroOne() - 0.5f) * 5.0f;
+        unkDir.z = 5.0f * Math_Coss(this->dyna.actor.posRot.rot.y + 0x8000) + (Math_Rand_ZeroOne() - 0.5f) * 5.0f;
+        unkDir.y = (Math_Rand_ZeroOne() - 0.5f) * 1.5f;
+        func_80028B18(globalCtx, &unkPos, &unkDir, &D_808865A8);
+    }
+}
+
+void BgHidanDalm_Update(Actor* thisx, GlobalContext* globalCtx) {
+    BgHidanDalm* this = THIS;
+
+    this->actionFunc(this, globalCtx);
+    Actor_MoveForward(thisx);
+    func_8002E4B4(globalCtx, thisx, 10.0f, 15.0f, 32.0f, 5);
```

Comment:
does using &this->actor work? we prefer that when possible so that we're not mixing this and thisx usage.
This is relevant for the other main functions in this file.
If its needed to match for any of them though then dont worry about it